File: curriculum-tools/courses/postgres-essentials/validation/author-23-24.ts
import { WAL } from "../curriculum/18-wal.ts";

const psql = Deno.env.get("PSQL") ?? "/usr/lib/postgresql/16/bin/psql";
const host = Deno.env.get("PGHOST") ?? "";
if (!host.startsWith("/tmp/pg-essentials-validation-author-23-24-")) {
  throw new Error("PGHOST must name an owned author-23-24 validation socket");
}

const modes = Deno.args.length ? Deno.args : ["23-core", "23-variation", "24-core", "24-variation"];

for (const mode of modes) {
  const match = mode.match(/^(23|24)-(core|variation)$/);
  if (!match) throw new Error(`unknown selection: ${mode}`);
  const lesson = WAL.lessons[Number(match[1]) - 23];
  let sql: string;
  if (match[2] === "core") {
    sql = `${lesson.setup ?? ""}\n${lesson.code}`;
  } else {
    const fenced = lesson.challenge?.match(/```sql\n([\s\S]*?)\n```/);
    if (!fenced) throw new Error(`missing SQL variation for lesson ${match[1]}`);
    sql = fenced[1];
  }

  const command = new Deno.Command(psql, {
    args: ["-X", "-v", "ON_ERROR_STOP=1", "-P", "pager=off"],
    stdin: "piped",
    stdout: "piped",
    stderr: "piped",
  }).spawn();
  const writer = command.stdin.getWriter();
  await writer.write(new TextEncoder().encode(sql));
  await writer.close();
  const result = await command.output();
  console.log(`=== ${mode} ===`);
  await Deno.stdout.write(result.stdout);
  await Deno.stderr.write(result.stderr);
  if (!result.success) Deno.exit(result.code);
}

File: curriculum-tools/courses/postgres-essentials/validation/author-23-24.py
#!/usr/bin/env python3
"""Run exact lesson 23–24 source in a disposable PostgreSQL 16 cluster."""

import os
import pwd
import shutil
import subprocess
import tempfile
from pathlib import Path

ROOT_PREFIX = "/tmp/pg-essentials-validation-author-23-24-"
PG_BIN = Path("/usr/lib/postgresql/16/bin")
PORT = "6554"


def run(command: list[str], *, env: dict[str, str] | None = None) -> None:
    subprocess.run(command, check=True, env=env, timeout=120)


def main() -> None:
    root = Path(tempfile.mkdtemp(prefix=Path(ROOT_PREFIX).name, dir="/tmp"))
    if not str(root).startswith(ROOT_PREFIX):
        raise RuntimeError(f"refusing unexpected fixture path: {root}")
    data = root / "data"
    socket = root / "socket"
    try:
        data.mkdir()
        socket.mkdir()
        postgres = pwd.getpwnam("postgres")
        for path in (root, data, socket):
            os.chown(path, postgres.pw_uid, postgres.pw_gid)
        run([
            "runuser", "-u", "postgres", "--", str(PG_BIN / "initdb"),
            "-D", str(data), "--no-locale", "--encoding=UTF8", "--auth=trust",
        ])
        run([
            "runuser", "-u", "postgres", "--", str(PG_BIN / "pg_ctl"),
            "-D", str(data), "-l", str(root / "server.log"), "-o",
            f"-k {socket} -p {PORT} -c listen_addresses= -c shared_buffers=16MB -c fsync=on",
            "start",
        ])
        env = {key: value for key, value in os.environ.items() if not key.startswith("PG")}
        env.update({
            "PGHOST": str(socket), "PGPORT": PORT,
            "PGUSER": "postgres", "PGDATABASE": "postgres",
        })
        run([
            "/root/.deno/bin/deno", "run", "--allow-env", "--allow-run",
            "courses/postgres-essentials/validation/author-23-24.ts",
        ], env=env)
    finally:
        if (data / "PG_VERSION").exists():
            status = subprocess.run([
                "runuser", "-u", "postgres", "--", str(PG_BIN / "pg_ctl"),
                "-D", str(data), "status",
            ], check=False, timeout=30, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            if status.returncode == 0:
                run([
                "runuser", "-u", "postgres", "--", str(PG_BIN / "pg_ctl"),
                "-D", str(data), "stop", "-m", "fast",
                ])
            stopped = subprocess.run([
                "runuser", "-u", "postgres", "--", str(PG_BIN / "pg_ctl"),
                "-D", str(data), "status",
            ], check=False, timeout=30, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            if stopped.returncode not in (3, 4) or (data / "postmaster.pid").exists():
                raise RuntimeError(f"cannot prove stopped; refusing to remove fixture: {root}")
        shutil.rmtree(root, ignore_errors=False)
        print(f"cleanup_removed={root}")


if __name__ == "__main__":
    main()
